backend/src/rules/victory_conditions.clj

```diff
@@ -17,42 +17,31 @@
 (defn player-wins-row?
   "Tells us if a player is winning a row"
   [game-state row-id player-id]
-  (let [opponent-id (if (= player-id (first (:player-ids game-state)))
-                      (second (:player-ids game-state))
-                      (first (:player-ids game-state)))]
+  (let [opponent-id (first (filter
+                             #(not= % player-id)
+                             (:player-ids game-state)))]
     (> (points-in-row game-state row-id player-id)
        (points-in-row game-state row-id opponent-id))))
 
 (defn get-won-rows
   "Tells us how many rows a player is winning"
   [game-state player-id]
-  (loop [won 0 row 0]
-    (if (>= row (count (:rows game-state)))
-      won
-      (recur
-        (if (player-wins-row? game-state row player-id)
-          (inc won)
-          won)
-        (inc row)))))
+  (let [row-num (count (:rows game-state))]
+    (count
+      (filter #(player-wins-row? game-state % player-id)
+              (range 0 (count (:rows game-state)))))))
 
 (defn most-won-rows
   "Which player is winning the most rows?"
   [game-state]
   (let [player-ids (:player-ids game-state)
         won-rows (map #(get-won-rows game-state %) player-ids)]
-    (loop [winner ""
-           most-wons 0
-           player-ids player-ids]
-      (if (empty? player-ids)
-        winner
-        (let [tmp-wons (get-won-rows game-state (first player-ids))]
-          (cond (= tmp-wons most-wons)
-                (recur "" most-wons (rest player-ids))
-                (> tmp-wons most-wons)
-                (recur (first player-ids) tmp-wons (rest player-ids))
-                :else
-                (recur winner most-wons (rest player-ids))))))))
-
+    (cond
+      (nil? player-ids) ""
```

Comment:
This is to pass a certain test case where there are no player-ids... I think we should just consider that an invalid input and return an `:error` though, hence why I put a special case here.

I reordered the comparisons for readability.